src/main/java/se/liu/ida/hefquin/engine/wrappers/graphqlwrapper/impl/SPARQL2GraphQLTranslatorImpl.java

```diff
@@ -0,0 +1,454 @@
+import se.liu.ida.hefquin.engine.wrappers.graphqlwrapper.query.impl.GraphQLQueryImpl;
+import se.liu.ida.hefquin.engine.wrappers.graphqlwrapper.utils.GraphCycleDetector;
+import se.liu.ida.hefquin.engine.wrappers.graphqlwrapper.utils.SGPNode;
+import se.liu.ida.hefquin.engine.wrappers.graphqlwrapper.utils.URI2GraphQLHelper;
+
+public class SPARQL2GraphQLTranslatorImpl implements SPARQL2GraphQLTranslator {
+
+    @Override
+    public GraphQLQuery translateBGP(final BGP bgp, final GraphQL2RDFConfiguration config,
+            final GraphQLEndpoint endpoint) {
+
+        // Initialize necessary data structures
+        final Map<Node, Set<TP>> subgraphPatterns = new HashMap<>();
+        final Map<Integer, Node> connectors = new HashMap<>();
+
+        // Partition BGP into SGPs and give unique integer id to TPs
+        final Set<? extends TriplePattern> bgpSet = bgp.getTriplePatterns();
+        int idCounter = 0;
+        for(final TriplePattern t : bgpSet){
+            final TP wrappedTriplePattern = new TP(idCounter,t);
+            final Node subject = t.asJenaTriple().getSubject();
+            if(subgraphPatterns.containsKey(subject)){
+                subgraphPatterns.get(subject).add(wrappedTriplePattern);
+            }
+            else{
+                final Set<TP> sgp = new HashSet<>();
+                sgp.add(wrappedTriplePattern);
+                subgraphPatterns.put(subject, sgp);
+            }
+            ++idCounter;
+        }
+
+        // Check for dependencies between the SGPs and create connectors
+        final Map<Node,SGPNode> sgpNodes = new HashMap<>();
+        for(final Node subject : subgraphPatterns.keySet()){
+            final Set<TP> sgp = subgraphPatterns.get(subject);
+            for(final TP tp : sgp){
+                final Node object = tp.getTriplePattern().asJenaTriple().getObject();
+                if(subgraphPatterns.containsKey(object) && ! object.equals(subject)){
+                    if(!sgpNodes.containsKey(subject)){
+                        sgpNodes.put(subject,new SGPNode());
+                    }
+                    if(!sgpNodes.containsKey(object)){
+                        sgpNodes.put(object,new SGPNode());
+                    }
+                    final SGPNode subjectSgpNode = sgpNodes.get(subject);
+                    final SGPNode objectSgpNode = sgpNodes.get(object);
+                    subjectSgpNode.addAdjacentNode(tp.getId(),objectSgpNode);
+                    connectors.put(tp.getId(), object);
+                }
+            }
+        }
+        // Remove all potential cyclic connector bindings
+        final Set<Integer> connectorsToBeRemoved = GraphCycleDetector.determineCyclicConnectors(sgpNodes);
+        for(final int i : connectorsToBeRemoved){
+            connectors.remove(i);
+        }
+
+        // Creating necessary variables for the GraphQL query
+        final TreeSet<String> fieldPaths = new TreeSet<>();
```

Comment:
Please use `Set` as the type. Also, the usual question if it is really necessary to keep a sorted map instead of a `HashSet`

Reply:
The reason `fieldPaths` is a TreeSet is because of how the toString method in the `GraphQLQueryImpl` works. Having it as a TreeSet ensures all fields of the same object are together in successive order which really facilitates an easy printing of the query. I could change it to a Set in all the files but then I would have to sort the Set into a TreeSet every time the toString method is used. What do you think?

Reply:
Maybe there is something I'm not seeing, but I don't think that it is worth it to keep a sorted set just to print the GraphQL queries nicely, unless it is a requirement (which I don't think it is).